src/rust/engine/process_execution/src/speculate.rs

```diff
@@ -0,0 +1,218 @@
+use super::{CommandRunner, ExecuteProcessRequest, FallibleExecuteProcessResult};
+use boxfuture::{BoxFuture, Boxable};
+use futures::future::{err, ok, Future};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio_timer::Delay;
+use workunit_store::WorkUnitStore;
+
+#[derive(Clone)]
+pub struct SpeculatingCommandRunner {
+  primary: Arc<dyn CommandRunner>,
+  secondary: Arc<dyn CommandRunner>,
+  speculation_timeout: Duration,
+}
+
+impl SpeculatingCommandRunner {
+  pub fn new(
+    primary: Box<dyn CommandRunner>,
+    secondary: Box<dyn CommandRunner>,
+    speculation_timeout: Duration,
+  ) -> SpeculatingCommandRunner {
+    SpeculatingCommandRunner {
+      primary: primary.into(),
+      secondary: secondary.into(),
+      speculation_timeout: speculation_timeout,
+    }
+  }
+}
+
+impl CommandRunner for SpeculatingCommandRunner {
+  fn run(
+    &self,
+    req: ExecuteProcessRequest,
+    workunit_store: WorkUnitStore,
+  ) -> BoxFuture<FallibleExecuteProcessResult, String> {
+    let command_runner = self.clone();
+    let workunit_store_clone = workunit_store.clone();
+    let req_2 = req.clone();
+    let delay = Delay::new(Instant::now() + self.speculation_timeout);
+    self
+      .primary
+      .run(req, workunit_store)
+      .select(delay.then(move |_| command_runner.secondary.run(req_2, workunit_store_clone)))
+      .then(|raced_result| match raced_result {
+        Ok((successful_res, _outstanding_req)) => {
+          ok::<FallibleExecuteProcessResult, String>(successful_res).to_boxed()
```

Comment:
Because `Result` implements `IntoFuture`, you should actually be able to just:
`Ok(successful_res)`
and
`Err(failed_res)`
in this closure :) Not particularly important, but good to know about :)